core/src/main/java/org/apache/calcite/rex/RexCorrelVariable.java

```diff
@@ -54,6 +54,20 @@
   @Override public SqlKind getKind() {
     return SqlKind.CORREL_VARIABLE;
   }
+
+  @Override public boolean equals(Object obj) {
+    if (obj == null || obj.getClass() != this.getClass()) {
+      return false;
+    }
+    RexCorrelVariable o = (RexCorrelVariable) obj;
+    return o == this || Objects.equals(digest, o.digest)
+        && Objects.equals(type, o.type)
+        && Objects.equals(id, o.id);
+  }
+
+  @Override public int hashCode() {
+    return Objects.hash(digest, type, id);
```

Comment:
is `digest` initialized by the time `hashCode` is executed?
Should `toString` be used here?

Reply:
I'm not sure...but we shouldn't use `node.toString()` everywhere when we are comparing RexNode-s...that's even worse...

I think the rational behind using digest instead of real tree comparison is to avoid the recursive comparision/etc - as long as when someone creates digest; remembers this design decision I think we are okay - but; for the record: I wouldn't expect this from a field like "digest"...